src/lang/lang.ml

```diff
@@ -715,3 +717,27 @@ let interactive () =
   in
   loop ();
   Tutils.shutdown ()
+
+(* Abstract types. *)
+
+module type Abstract = sig
+  type content
+
+  val t : t
+  val to_value : content -> value
+  val of_value : value -> content
+end
+
+module type AbstractDef = sig
+  type content
+
+  val name : string
+end
+
+module MkAbstract (Def : AbstractDef) = struct
+  let t = ground_t (T.Abstract Def.name)
+  let to_value c = mk ~t:string_t (Abstract (Def.name, Obj.magic c))
```

Comment:
Can't we avoid `Obj.magic` if we have a static list of abstract types, e.g. with a GADT?